src/System.CommandLine/Symbol.cs

```diff
@@ -59,22 +59,37 @@ private protected void AddArgumentInner(Argument argument)
 
         public virtual IEnumerable<string?> GetSuggestions(ParseResult? parseResult = null, string? textToMatch = null)
         {
-            var argumentSuggestions =
-                Children
-                    .OfType<IArgument>()
-                    .SelectMany(a => a.GetSuggestions(parseResult, textToMatch))
-                    .ToArray();
-
-            return Children
-                   .Where(s => !s.IsHidden)
-                   .OfType<IIdentifierSymbol>()
-                   .SelectMany(s => s.Aliases)
-                   .Concat(argumentSuggestions)
-                   .Distinct()
-                   .Containing(textToMatch)
-                   .Where(symbol => symbol != null)
-                   .OrderBy(symbol => symbol!.IndexOfCaseInsensitive(textToMatch))
-                   .ThenBy(symbol => symbol, StringComparer.OrdinalIgnoreCase);
+            var suggestions = new HashSet<string>();
+
+            foreach (var child in Children)
+            {
+                switch (child)
+                {
+                    case IIdentifierSymbol identifier when !child.IsHidden:
+                        foreach (var alias in identifier.Aliases)
+                        {
+                            if (alias is string s && s.ContainsCaseInsensitive(textToMatch))
+                            {
+                                suggestions.Add(s);
+                            }
+                        }
+                        break;
+                    case IArgument argument:
+                        foreach (var suggestion in argument.GetSuggestions(parseResult, textToMatch))
+                        {
+                            if (suggestion is string s && s.ContainsCaseInsensitive(textToMatch))
+                            {
+                                suggestions.Add(s);
+                            }
+                        }
+                        break;
+                }
+            }
+
+            return suggestions
+                .OrderBy(symbol => symbol!.IndexOfCaseInsensitive(textToMatch))
+                .ThenBy(symbol => symbol, StringComparer.OrdinalIgnoreCase)
+                .ToList();
```

Comment:
Is pre-emptively calling `ToList` useful here?

Reply:
Not if it's iterated only once, which it seems to be the case. I'll remove it.